src/cargo/ops/cargo_package.rs

```diff
@@ -352,6 +353,18 @@ fn run_verify(ws: &Workspace, tar: &FileLock, opts: &PackageOpts) -> CargoResult
         Arc::new(DefaultExecutor),
     )?;
 
+    // Check that build.rs didn't modify any files in the src directory.
+    let ws_fingerprint = src.last_modified_file(ws.current()?)?;
+    if pkg_fingerprint != ws_fingerprint {
+        let (_, path) = ws_fingerprint;
+        bail!(
+            "Source directory was modified by build.rs during cargo publish. \
+             Build scripts should not modify anything outside of OUT_DIR. \
+             Modified file: {}",
```

Comment:
One more thing: let's add `use '--no-verify' flag to suppress this check` to the error message?